invoices/invoiceregistry.go

```diff
@@ -781,6 +781,72 @@ func (i *InvoiceRegistry) processKeySend(ctx invoiceUpdateCtx) error {
 	return nil
 }
 
+// processAMP just-in-time inserts an invoice if this htlc is a keysend
+// htlc.
+func (i *InvoiceRegistry) processAMP(ctx invoiceUpdateCtx) error {
+	// Only allow keysend for non-mpp payments.
+	if ctx.mpp == nil {
+		return errors.New("no MPP record for AMP")
+	}
+
+	// Create an invoice for the total amount expected, provided in the MPP
+	// record.
+	amt := ctx.mpp.TotalMsat()
+
+	// Set the TLV and MPP optional features on the invoice. We'll also make
+	// the AMP features required so that it can't be paid by legacy or MPP
+	// htlcs.
+	rawFeatures := lnwire.NewRawFeatureVector(
+		lnwire.TLVOnionPayloadOptional,
+		lnwire.PaymentAddrOptional,
+		lnwire.AMPRequired,
+	)
+	features := lnwire.NewFeatureVector(rawFeatures, lnwire.Features)
+
+	// Use the minimum block delta that we require for settling htlcs.
+	finalCltvDelta := i.cfg.FinalCltvRejectDelta
+
+	// Pre-check expiry here to prevent inserting an invoice that will not
+	// be settled.
+	if ctx.expiry < uint32(ctx.currentHeight+finalCltvDelta) {
+		return errors.New("final expiry too soon")
+	}
+
+	// We'll use the sender-generated payment address provided in the HTLC
+	// to create our AMP invoice.
+	payAddr := ctx.mpp.PaymentAddr()
+
+	// Create placeholder invoice.
+	invoice := &channeldb.Invoice{
+		CreationDate: i.cfg.Clock.Now(),
+		Terms: channeldb.ContractTerm{
+			FinalCltvDelta:  finalCltvDelta,
+			Value:           amt,
+			PaymentPreimage: nil,
+			PaymentAddr:     payAddr,
+			Features:        features,
+		},
+	}
+
+	if i.cfg.KeysendHoldTime != 0 {
```

Comment:
IIUC, as is, if this flag is set then we won't immeidatley settle things after all teh shards arrive. However this is only useful if the caller is able to manually settle the payment using the invoice RPC APIs. The existing settle invoice API is based on the caller knowing the pre-image (sub-swaps or they made it themselves) while in this case, such an action isn't possible as the code will attempt to look up the invoice via it's payment hash as a ref: https://github.com/lightningnetwork/lnd/blob/master/invoices/invoiceregistry.go#L1030

So I think we can remove this clause all together since it can't properly be used w/ the existing API. This was originally added for hodl keysend, but in that case there's a single pre-image that's returned via the subscription API once the invoice gets "just in time" inserted into the database.

Reply:
sounds good, we can add it back one we have the ability to settle cancel the set via RPC